File: migalha/Pattern.java
import java.util.*;

// defines the structure of a vulnerability: what counts as a source, sanitizer, and sink
public class Pattern {
    private String name;
    private Set<String> sources;
    private Set<String> sanitizers;
    private Set<String> sinks;

    public Pattern(String name, List<String> sources, List<String> sanitizers, List<String> sinks) {
        this.name = name;
        this.sources = new HashSet<>(sources);
        this.sanitizers = new HashSet<>(sanitizers);
        this.sinks = new HashSet<>(sinks);
    }

    public String getName() {
        return name;
    }

    public Set<String> getSources() {
        return new HashSet<>(sources);
    }

    public Set<String> getSanitizers() {
        return new HashSet<>(sanitizers);
    }

    public Set<String> getSinks() {
        return new HashSet<>(sinks);
    }

    public boolean isSource(String s) {
        return sources.contains(s);
    }

    public boolean isSanitizer(String s) {
        return sanitizers.contains(s);
    }

    public boolean isSink(String s) {
        return sinks.contains(s);
    }
}

File: migalha/Pattern.py
from typing import List, Set

class Pattern:
    """
    Defines the structure of a vulnerability: what counts as a source,
    sanitizer, and sink.
    """

    def __init__(self, name: str, sources: List[str], sanitizers: List[str], sinks: List[str]):
        self._name = name
        self._sources: Set[str] = set(sources)
        self._sanitizers: Set[str] = set(sanitizers)
        self._sinks: Set[str] = set(sinks)

    def get_name(self) -> str:
        return self._name

    def get_sources(self) -> Set[str]:
        return set(self._sources)

    def get_sanitizers(self) -> Set[str]:
        return set(self._sanitizers)

    def get_sinks(self) -> Set[str]:
        return set(self._sinks)

    def is_source(self, s: str) -> bool:
        return s in self._sources

    def is_sanitizer(self, s: str) -> bool:
        return s in self._sanitizers

    def is_sink(self, s: str) -> bool:
        return s in self._sinks

    def __repr__(self):
        return (f"Pattern(name={self._name!r}, "
                f"sources={self._sources}, "
                f"sanitizers={self._sanitizers}, "
                f"sinks={self._sinks})")
